src/foam/nanos/auth/CachingAuthService.java

```diff
@@ -66,7 +68,16 @@ public void eof() {
         public void reset(Detachable sub) {
           purgeCache(x);
         }
-      }, TRUE);
+      };
+
+      DAO userDAO       = (DAO) x.get("userDAO");
+      DAO groupDAO      = (DAO) x.get("groupDAO");
+      DAO groupPermissionJunctionDAO = (DAO) x.get("groupPermissionJunctionDAO");
+
+      groupDAO.listen(purgeSink, TRUE);
```

Comment:
We're setting up these listeners inside `getPermissionMap`, which means we're going to be setting up three listeners for every user in the system. Would it be more efficient to set up 3 listeners when this class is initialized and purge all caches in all sessions when an event happens? Genuine question. I'm not sure if the overhead of all of the listeners would be greater than or less than the overhead of doing the lookup of all sessions.

Reply:
If we just had three listeners then each session would have to listen for whatever those listeners trigger, which would result in the same amount of work.